core/src/main/scala/org/http4s/jdkhttpclient/JdkWSClient.scala

```diff
@@ -47,115 +48,141 @@ object JdkWSClient {
+                wsListener = new JWebSocket.Listener {
+                  override def onOpen(webSocket: JWebSocket): Unit = ()
+                  override def onClose(webSocket: JWebSocket, statusCode: Int, reason: String)
+                      : CompletionStage[_] =
+                    // The output side of this connection will be closed when the returned CompletionStage completes.
+                    // Therefore, we return a never completing CompletionStage, so we can control when the output will
+                    // be closed (as it is allowed to continue sending frames (as few as possible) after a close frame
+                    // has been received).
+                    handleReceive(WSFrame.Close(statusCode, reason).asRight)
+                      .thenCompose[Nothing](_ => new CompletableFuture[Nothing])
+                  override def onText(webSocket: JWebSocket, data: CharSequence, last: Boolean)
+                      : CompletionStage[_] =
+                    handleReceive(WSFrame.Text(data.toString, last).asRight)
+                  override def onBinary(webSocket: JWebSocket, data: ByteBuffer, last: Boolean)
+                      : CompletionStage[_] =
+                    handleReceive(WSFrame.Binary(ByteVector(data), last).asRight)
+                  override def onPing(webSocket: JWebSocket, message: ByteBuffer)
+                      : CompletionStage[_] =
+                    handleReceive(WSFrame.Ping(ByteVector(message)).asRight)
+                  override def onPong(webSocket: JWebSocket, message: ByteBuffer)
+                      : CompletionStage[_] =
+                    handleReceive(WSFrame.Pong(ByteVector(message)).asRight)
+                  override def onError(webSocket: JWebSocket, error: Throwable): Unit = {
+                    handleReceive(error.asLeft); ()
+                  }
                 }
-                builder
-              }
-              queue <- Queue.unbounded[F, Either[Throwable, WSFrame]]
-              closedDef <- Deferred[F, Unit]
-              handleReceive =
-                (wsf: Either[Throwable, WSFrame]) =>
-                  dispatcher.unsafeToCompletableFuture(
-                    queue.offer(wsf) *> (wsf match {
-                      case Left(_) | Right(_: WSFrame.Close) => closedDef.complete(()).void
-                      case _ => F.unit
-                    })
+                webSocket <- F.fromCompletableFuture(
+                  F.delay(wsBuilder.buildAsync(URI.create(req.uri.renderString), wsListener))
+                )
+                sendSem <- Semaphore[F](1L)
+              } yield (webSocket, queue, closedDef, sendSem)
+            } { case (webSocket, queue, closedDef, sendSem) =>
+              def awaitClose(step: Int): F[Unit] =
+                closedDef.tryGet.flatMap {
+                  case Some(_) => F.unit
+                  case None =>
+                    if (step < 10) F.sleep(100.millis) *> awaitClose(step + 1) else F.unit
```

Comment:
I wonder what a sensible amount of time to wait here is? should this be configurable? overloaded `apply` with a duration parameter to retain backwards compat? In practice this alleviates the need for actually using `abort` quite a bit which is nice.

Reply:
I'm a bit confused by this, is this equivalent to `closedDef.get.timeout(1.second)` ?

Reply:
that's what I was looking for. For some reason it wasn't in my code completion :facepalm: . thanks

Reply:
Ah yeah, you probably just needed `import cats.effect.syntax.all._` to get that :)

Sorry, another follow-up question to this: do _we_ have to do this timeout? I am wondering if the JDK WebSocket implementation itself has an internal timeout, and we are just forgetting to listen for an error event or something. see the docs:

> Unless the CompletableFuture returned from this method completes with IllegalArgumentException, or the method throws NullPointerException, the output will be closed.
> 
> If not already closed, the input remains open until a Close message [received](https://docs.oracle.com/en/java/javase/17/docs/api/java.net.http/java/net/http/WebSocket.Listener.html#onClose(java.net.http.WebSocket,int,java.lang.String)), or abort is invoked, or an [error](https://docs.oracle.com/en/java/javase/17/docs/api/java.net.http/java/net/http/WebSocket.Listener.html#onError(java.net.http.WebSocket,java.lang.Throwable)) occurs.

https://docs.oracle.com/en/java/javase/17/docs/api/java.net.http/java/net/http/WebSocket.html#sendClose(int,java.lang.String)